demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,448 @@
+                               uint16_t nameIndex,
+                               uint16_t filterIndex );
+
+/**
+ * @brief Attempt to match characters in a topic filter by wildcards.
+ *
+ * @param[in] pTopicFilter The topic filter containing the wildcard.
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] filterIndex Index of the wildcard in the topic filter.
+ * @param[in,out] pNameIndex Index of character in topic name. This variable is
+ * advanced for `+` wildcards.
+ * @param[out] pMatch Whether the topic filter and topic name match.
+ *
+ * @return `true` if the caller of this function should exit; `false` if the caller
+ * should continue parsing the topics.
+ */
+static bool matchWildcards( const char * pTopicFilter,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t filterIndex,
+                            uint16_t * pNameIndex,
+                            bool * pMatch );
+
+/**
+ * @brief Match a topic name and topic filter allowing the use of wildcards.
+ *
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] pTopicFilter The topic filter to check.
+ * @param[in] topicFilterLength Length of topic filter.
+ *
+ * @return `true` if the topic name and topic filter match; `false` otherwise.
+ */
+static bool topicFilterMatch( const char * pTopicName,
+                              uint16_t topicNameLength,
+                              const char * pTopicFilter,
+                              uint16_t topicFilterLength );
+
+/**
+ * @brief Matches a topic name (from a incoming PUBLISH) with a topic filter.
+ *
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] pTopicFilter The topic filter to check.
+ * @param[in] topicFilterLength Length of topic filter.
+ *
+ * @return `true` if the topic name and topic filter match; `false`
+ * otherwise.
+ */
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength );
+
+/*-----------------------------------------------------------*/
+
+static bool matchEndWildcards( const char * pTopicFilter,
+                               uint16_t topicFilterLength,
+                               uint16_t topicNameLength,
+                               uint16_t nameIndex,
+                               uint16_t filterIndex )
+{
+    bool matchFound = false, endChar = false;
+
+    /* Determine if the last character is reached for the topic name, and the
+     * second last character for the topic filter to match for the '#' wildcard. */
+    endChar = ( nameIndex == ( topicNameLength - 1U ) ) &&
```

Comment:
The name `endChar` seems a bit misleading.

Reply:
Couldn't find an appropriate name, so have removed the use of `endChar` altogether to instead have the conditional expression at `if`